localstack-core/localstack/utils/lambda_debug_mode/lambda_debug_mode_config.py

```diff
@@ -0,0 +1,198 @@
+    port_number: int
+
+    def __init__(self, port_number: int):
+        self.port_number = port_number
+
+    def __str__(self):
+        return f"PortAlreadyInUse: '{self.port_number}'"
+
+
+class DuplicateLambdaDebugConfig(LambdaDebugModeConfigException):
+    lambda_arn_debug_config_first: str
+    lambda_arn_debug_config_second: str
+
+    def __init__(self, lambda_arn_debug_config_first: str, lambda_arn_debug_config_second: str):
+        self.lambda_arn_debug_config_first = lambda_arn_debug_config_first
+        self.lambda_arn_debug_config_second = lambda_arn_debug_config_second
+
+    def __str__(self):
+        return (
+            f"DuplicateLambdaDebugConfig: Lambda debug configuration in '{self.lambda_arn_debug_config_first}' "
+            f"is redefined in '{self.lambda_arn_debug_config_second}'"
+        )
+
+
+class _LambdaDebugModeConfigValidationState:
+    ports_used: set[int]
+
+    def __init__(self):
+        self.ports_used = set()
+
+
+class _SafeLoaderWithDuplicateCheck(SafeLoader):
+    def __init__(self, stream):
+        super().__init__(stream)
+        self.add_constructor(
+            yaml.resolver.BaseResolver.DEFAULT_MAPPING_TAG,
+            self._construct_mappings_with_duplicate_check,
+        )
+
+    @staticmethod
+    def _construct_mappings_with_duplicate_check(loader: Loader, node: MappingNode, deep=False):
+        # Constructs yaml bindings, whilst checking for duplicate mapping key definitions, raising a
+        # MarkedYAMLError when one is found.
+        mapping = dict()
+        for key_node, value_node in node.value:
+            key = loader.construct_object(key_node, deep=deep)
+            if key in mapping:
+                # Create a MarkedYAMLError to indicate the duplicate key issue
+                raise MarkedYAMLError(
+                    context="while constructing a mapping",
+                    context_mark=node.start_mark,
+                    problem=f"found duplicate key: {key}",
+                    problem_mark=key_node.start_mark,
+                )
+            value = loader.construct_object(value_node, deep=deep)
+            mapping[key] = value
+        return mapping
+
+
+def from_yaml_string(yaml_string: str) -> Optional[LambdaDebugModeConfig]:
+    try:
+        data = yaml.load(yaml_string, _SafeLoaderWithDuplicateCheck)
+    except yaml.YAMLError as yaml_error:
+        LOG.error(
+            f"Could not parse yaml lambda debug mode configuration file due to: {str(yaml_error)}"
+        )
+        data = None
+    if not data:
+        return None
+    config = LambdaDebugModeConfig(**data)
+    return config
+
+
+def validate_lambda_debug_mode_config(config: LambdaDebugModeConfig) -> None:
+    _validate_lambda_debug_mode_config(
+        validation_state=_LambdaDebugModeConfigValidationState(), config=config
+    )
+
+
+def _validate_lambda_debug_mode_config(
+    validation_state: _LambdaDebugModeConfigValidationState, config: LambdaDebugModeConfig
+):
```

Comment:
This function is called `validate_` but does actively change the values of the config (by replacing unqualified ARNs with qualified ARNs). It should be clear from the method that we are not only validating here, but also "canonicalizing". So maybe renaming it, or at least adding it in a docstring or something.